src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -4520,6 +4520,20 @@ class RGWBucketSyncSingleEntryCR : public RGWCoroutine {
             tn->log(0, "entry with empty obj name, skipping");
             goto done;
           }
+
+          // make sure versioned object only lands on versioned bucket and non-versioned object only lands on non-versioned bucket
+          if (key.instance.empty() == sync_pipe.dest_bucket_info.versioned()) {
```

Comment:
I wonder whether we need to consider versioning suspended here.

Reply:
two tests failed here: https://tracker.ceph.com/issues/70728.
`test_null_version_id_delete()` failed because the check disallows object deletes to be synced when a previously null-versioned object is deleted on source. `test_version_suspended_incremental_sync()` fails because we don't consider versioning suspended as Seena pointed out. I have raised https://github.com/ceph/ceph/pull/62594 to address that.